File: hw4/matlab/filters.m
img=imread('HW4_1.tif'); %we are reading image in here
figure
imshow(img)
% [a,b,c]=size(img);
% f=8;
% filter=ones(f,f);
% filter=(filter)/sum(filter(:));
% filter=double(filter);
% array=padarray(img,[6 6]);
% array=double(array);
% array1=array(:,:,1);
% array2=array(:,:,2);
% array3=array(:,:,3);
% output1=zeros(a,b);
% output2=zeros(a,b);
% output3=zeros(a,b);
% for i=1:size(array,1)-f-1
%     for j=1:size(array,2)-f-1
%     temp1=array1(i:i+f-1,j:j+f-1).*filter;
%     output1(i,j)=sum(temp1(:));
%     temp2=array2(i:i+f-1,j:j+f-1).*filter;
%     output2(i,j)=sum(temp2(:));
%     temp3=array3(i:i+f-1,j:j+f-1).*filter;
%     output3(i,j)=sum(temp3(:));
%     end
% end
% 
%  im=cat(3,output1,output2,output3);
%  figure
%  imshow(uint8(im))
%  disp(im)

[a,b]=size(img);
%Mean Filter
% f=3;
% filter=ones(f,f);
% filter=(filter)/sum(filter(:));
% filter=double(filter);
% array=padarray(img,[1 1]);
% array=double(array);
% output1=zeros(a,b);
% for i=1:size(array,1)-f-1
%     for j=1:size(array,2)-f-1
%     temp1=array1(i:i+f-1,j:j+f-1).*filter;
%     output1(i,j)=sum(temp1(:));
%     end
% end
%   figure
% imshow(uint8(output1))
%------------------------------------------------------------------------
%median filter
mf=ones(3,3);
output=zeros(size(img)-2);
for i=1:size(img,1)-2
    for j=1:size(img,2)-2
    temp1=double(img(i:i+2,j:j+2)).*mf;
    output(i,j)=median(temp1(:));
    end
end

imshow(uint8(output))
%------------------------------------------------------------------------
%max filter
A=img;
B=zeros(size(A));
modifyA=padarray(A,[1 1]);
x=[1:3]';
y=[1:3]'; 
for i= 1:size(modifyA,1)-2
    for j=1:size(modifyA,2)-2
       window=reshape(modifyA(i+x-1,j+y-1),[],1);
       B(i,j)=max(window);
    end
end
B=uint8(B);
figure
imshow(B)
%------------------------------------------------------------------------
%min filter
A=img;
B=zeros(size(A));
modifyA=padarray(A,[1 1]);
x=[1:3]';
y=[1:3]'; 
for i= 1:size(modifyA,1)-2
    for j=1:size(modifyA,2)-2
       window=reshape(modifyA(i+x-1,j+y-1),[],1);
       B(i,j)=min(window);
    end
end
B=uint8(B);
figure
imshow(B)

File: hw4/python/filters.py
import cv2
import numpy as np
import statistics
img=cv2.imread("imagess/HW4_1.tif")
cv2.imshow("original window",img)
img=cv2.cvtColor(img,cv2.COLOR_BGR2GRAY)
[a,b]=img.shape
array=np.zeros(shape=(a+2,b+2))
array2=np.zeros(shape=(a+2,b+2))
array3=np.zeros(shape=(a+2,b+2))
array4=np.zeros(shape=(a+2,b+2))
for i in range(a):
    for j in range(b):
        x=img[i,j]
        array[i+1,j+1]=x
        array2[i + 1, j + 1] = x
        array3[i + 1, j + 1] = x
        array4[i + 1, j + 1] = x


#mean filtering
filter=np.ones([3,3])
filter=(filter)/9
filter=np.double(filter)
array=np.double(array)
output1=np.zeros([a,b])
for i in range(a-2):
     for j in range(b-2):
         temp1=np.multiply(array[i:i+3,j:j+3],filter)
         output1[i,j]=temp1.sum()



image=np.uint8(output1)
cv2.imshow("image window",image)

#median filtering
filter2=np.ones([3,3])
filter2=np.double(filter2)
array2=np.double(array2)
output2=np.zeros([a,b])
output3=np.zeros([a,b])
output4=np.zeros([a,b])
for i in range(a-2):
     for j in range(b-2):
         temp1=np.multiply(array2[i:i+3,j:j+3],filter2)
         output2[i,j]=np.median(temp1)


image=np.uint8(output2)
cv2.imshow("image 2 window",image)



for i in range(a-2):
     for j in range(b-2):
         temp1=array3[i:i+3,j:j+3]
         output3[i,j]=np.max(temp1)

image=np.uint8(output3)
cv2.imshow("image 3 window",image)

for i in range(a-2):
     for j in range(b-2):
         temp1=array4[i:i+3,j:j+3]
         output4[i,j]=np.min(temp1)

image=np.uint8(output4)
cv2.imshow("image 4 window",image)


cv2.waitKey(0)
cv2.destroyAllWindows()